flytekit/core/external_api_task.py

```diff
@@ -0,0 +1,73 @@
+import collections
+import inspect
+from abc import abstractmethod
+from typing import Any, Dict, Optional, TypeVar
+
+import jsonpickle
+from flyteidl.admin.agent_pb2 import DoTaskResponse
+from typing_extensions import get_type_hints
+
+from flytekit.configuration import SerializationSettings
+from flytekit.core.base_task import PythonTask
+from flytekit.core.interface import Interface
+from flytekit.extend.backend.base_agent import SyncAgentExecutorMixin
+
+T = TypeVar("T")
+TASK_MODULE = "task_module"
+TASK_NAME = "task_name"
+TASK_CONFIG_PKL = "task_config_pkl"
+TASK_TYPE = "api_task"
+
+
+class ExternalApiTask(SyncAgentExecutorMixin, PythonTask):
+    """
+    Base class for all external API tasks. External API tasks are tasks that are designed to run until they receive a
+    response from an external service. When the response is received, the task will complete. External API tasks are
+    designed to be run by the flyte agent.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        config: Optional[T] = None,
+        task_type: str = TASK_TYPE,
+        return_type: Optional[Any] = None,
+        **kwargs,
+    ):
+        type_hints = get_type_hints(self.do, include_extras=True)
+        signature = inspect.signature(self.do)
+        inputs = collections.OrderedDict()
+        outputs = collections.OrderedDict({"o0": return_type}) if return_type else collections.OrderedDict()
+
+        for k, _ in signature.parameters.items():  # type: ignore
+            annotation = type_hints.get(k, None)
+            inputs[k] = annotation
+
+        super().__init__(
+            task_type=task_type,
+            name=name,
+            task_config=config,
+            interface=Interface(inputs=inputs, outputs=outputs),
+            is_sync_plugin=True,
+            **kwargs,
+        )
+
+        self._config = config
+
+    @abstractmethod
+    async def do(self, **kwargs) -> DoTaskResponse:
+        """
+        Initiate an HTTP request to an external service such as OpenAI or Vertex AI and retrieve the response.
+        """
+        raise NotImplementedError
+
+    def get_custom(self, settings: SerializationSettings = None) -> Dict[str, Any]:
```

Comment:
why do we have to use pickle?
cc @pingsutw do you know? is this to maintain simply python coversion? This is potentially dangerous as it may break across python versions.

Reply:
pickle is faster, we've talked about this before and Kevin told me that it might be ok.
Will there be multiple python versions usecases?

Reply:
we cannot control backend and flytekit version

Reply:
you are right, I fix it!
Thank you very much for your advice!